Breadcrumbs integration

vitvakatu opened the pull request:
### Pull Request Description

[ci no changelog needed]
[Task link](https://www.pivotaltracker.com/story/show/182675703)

This PR implements the actual integration of the breadcrumbs panel with the component list panel. A special breadcrumbs controller (`controller::searcher::breadcrumbs`) is tracking the currently opened module with a list of its parents. The searcher presenter uses the API of the controller to sync the displayed list of breadcrumbs with the controller state.


https://user-images.githubusercontent.com/6566674/193064122-7d3fc4d6-9148-4ded-a73e-767ac9ac83f8.mp4



### Important Notes

- There is an `All` breadcrumb displayed at all times at the beginning of the list. It will be replaced with a section name as part of [Section Title on Component Browser's Breadcrumbs Panel](https://www.pivotaltracker.com/story/show/182610561) task.
- I changed the implementation of `project::main_module_id`, `project::QualifiedName::main_module`, and `API::main_module` so that they are logically connected which each other.
- I adjusted the Breadcrumbs View to avoid "appearance" animation glitches when opening new modules. `set_entries` was replaced with the `set_entries_from` endpoint.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

vitvakatu committed:
Implement entering modules in component browser

vitvakatu committed:
Self-review

vitvakatu committed:
Enso-formatter

vitvakatu committed:
Clippy fixes

vitvakatu committed:
Integrate breadcrumbs

vitvakatu committed:
Address review comments

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
refactorings

vitvakatu committed:
refactorings

vitvakatu committed:
Lazy update of the displayed breadcrumbs

vitvakatu committed:
Fix clippy issues

vitvakatu committed:
Enso formatter

vitvakatu committed:
Fix Breadcrumbs::set_content

vitvakatu committed:
Hide/show ellipsis

farmaazon commented on app/gui/controller/double-representation/src/project.rs:
```diff
@@ -23,7 +23,17 @@ pub const BASE_LIBRARY_NAME: &str = "Base";
+///
+/// Currently, this name is hardcoded in the engine services and is populated for each project
+/// created using engine's Project Picker service.
+pub const INITIAL_MODULE_NAME: &str = "Main";
```
I'm not sure this constant should be here; the Double Representation crate is mailny for synchronizing text and graph representations of our code.

vitvakatu replied:
Removed it and used the constant from the AST module instead.

farmaazon commented on app/gui/src/controller/searcher/component/builder.rs:
```diff
@@ -49,26 +49,39 @@ pub struct ModuleGroups {
     pub flattened_content: Option<component::Group>,
     pub submodules:        component::group::AlphabeticalListBuilder,
     pub is_top_module:     bool,
+    pub qualified_name:    module::QualifiedName,
```
Let's put it as the first field.

farmaazon commented on app/gui/src/controller/searcher/component/builder.rs:
```diff
@@ -49,26 +49,39 @@ pub struct ModuleGroups {
     ///
     /// The existence of flattened content is decided during construction.
-    pub fn new(component_id: component::Id, entry: &suggestion_database::Entry) -> Self {
+    ///
+    /// Returns [`None`] if entry's qualified name can't be converted into a module name.
+    pub fn new(
+        component_id: component::Id,
+        entry: &suggestion_database::Entry,
+    ) -> FallibleResult<Self> {
         let is_top_module = entry.module.is_top_module();
```
The docs should say about `Err`, not `None`.

farmaazon commented on app/gui/src/controller/searcher/breadcrumbs.rs:
```diff
@@ -4,20 +4,24 @@ use crate::prelude::*;
-///
-/// TODO: The actual implementation would be finished in
-///   [Breadcrumbs Panel integration task](https://www.pivotaltracker.com/story/show/182675703).
+/// functionality of the breadcrumbs panel. It is used to store the state of the breadcrumbs
+/// panel, but it does not provide any view-related functionality. The integration between the
+/// controller and the view is done by the [searcher presenter](crate::presenter::searcher).
```
I think we do not need to mention that the controller does not provide any view-related functionality.

farmaazon commented on app/gui/src/controller/searcher/breadcrumbs.rs:
```diff
@@ -26,18 +30,145 @@ impl Breadcrumbs {
+    ///
+    /// Returns [`None`] if the [`module`] is not found in the database or in the components list.
+    pub fn build(self, module: &component::Id) -> Option<Vec<BreadcrumbEntry>> {
+        let mut result = Vec::new();
+        let module_name = self.components.module_qualified_name(*module)?;
+        let entry = BreadcrumbEntry::from(self.lookup(&module_name)?);
+        result.push(entry);
+        result.extend(self.collect_parents(&module_name));
+        let project_name = module_name.project_name.clone();
+        let main_module = module::QualifiedName::new_main(project_name.clone());
+        if let Some(entry) = self.lookup(&main_module) {
+            result.push(BreadcrumbEntry {
+                displayed_name: String::from(project_name.project).into(),
+                ..entry.into()
+            });
+        }
+        Some(result.reversed())
```
I strongly opt for avoid mutable vars where possible. In this case we can chain iterators `once(entry)` with result of (`sellf.collect_parentr`) with main module entry (an option is also an iterator). Then, collect the vector from the chain iterator.

farmaazon commented on app/gui/src/controller/searcher/breadcrumbs.rs:
```diff
@@ -26,18 +30,145 @@ impl Breadcrumbs {
+    /// Collect all parent modules of the given module.
+    ///
+    /// Panics if the module is not found in the database.
+    fn collect_parents(&self, name: &module::QualifiedName) -> Vec<BreadcrumbEntry> {
+        let mut result = Vec::new();
+        let (_, mut current) = self.lookup(name).expect("Entry should be present in the database.");
+        while let Some(parent) = current.parent_module() {
+            if let Some((component_id, parent_entry)) = self.lookup(&parent) {
+                result.push((component_id, parent_entry.clone_ref()).into());
+                current = parent_entry;
+            } else {
+                break;
+            }
+        }
+        result
```
We can make an iterator over all parent modules for `module::QualifiedName`. Then, this function could return `impl Iterator<Item=BreadcrumbEntry>` instead of Vec, thus avoiding allocation.

vitvakatu replied:
Refactored, but returning iterator is not possible here. We need to collect parents in reverse order, which is only possible for `DoubleEndedIterator`. As `module::QualifiedName::parent_modules` also returns `impl Iterator`, and it does not implement `DoubleEndedIterator`, we need to allocate a vector here.

farmaazon commented on app/gui/src/presenter/searcher.rs:
```diff
@@ -190,24 +191,43 @@ impl Model {
+
+    fn set_breadcrumbs(&self, names: impl Iterator<Item = ImString>) {
+        if let SearcherVariant::ComponentBrowser(browser) = self.view.searcher() {
+            browser.model().list.set_breadcrumbs_from((names.map(Into::into).collect(), 1));
```
What is this `1`?

vitvakatu replied:
The index to update the breadcrumbs from. Refactored and added a comment

farmaazon commented on lib/rust/ensogl/component/grid-view/src/column_widths.rs:
```diff
@@ -76,10 +76,11 @@ impl ColumnWidths {
     /// [`Self::resize`].
     pub fn pos_offset(&self, column: usize) -> f32 {
-        if column == 0 {
+        let borrowed = self.width_diffs.borrow();
+        if column == 0 || column > borrowed.len() {
```
Is this a normal routine that someone reads position of non-existing column? If not, the error or warning should be printed.

vitvakatu replied:
I did notice a panic in this place once. I'm not sure if this is a normal routine, so I left a warning here.

vitvakatu committed:
Addressing review comments

vitvakatu committed:
Revert if condition for readability

farmaazon commented on app/gui/src/ide.rs:
```diff
@@ -104,8 +103,6 @@ pub struct FailedIde {
-    model::module::Path::from_name_segments(project.project_content_root_id(), &[
-        INITIAL_MODULE_NAME,
-    ])
+pub async fn initial_module_path(project: &model::Project) -> FallibleResult<model::module::Path> {
+    Ok(project.main_module_model().await?.path().clone_ref())
```
Getting a module from the project is a quite big deal: That may mean opening the module in the engine, and setting up the edit updates. That's why main_module_model is asynchronous. I think we don't want to do run the entire process only to get the path of the module.

vitvakatu replied:
Fixed

farmaazon commented on app/gui/controller/double-representation/src/module.rs:
```diff
@@ -300,9 +298,24 @@ impl QualifiedName {
-        let project_name = self.project_name.clone();
-        Some(Self { project_name, id })
+        if self.is_top_module() {
+            None
```
This is something I'm not sure how to handle. Technically, one could say that the parent of the top module (like local::Project::Module) is actually the main module (local::Project). But I'm not sure how do we rely on this behavior somewhere else? On the other hand, maybe it would fix the "Project" crumb in Component Browser (where is no content yet)?

farmaazon reviewed (changes requested):
The comment regarding "parent_module" is for thought and discussion and not blocker.

However, making getter for initial module path asynchronous is.

vitvakatu committed:
Fix initial_module_path function

vitvakatu committed:
Fix "project" breadcrumb

vitvakatu commented:
After some experiments, I fixed the "Project" breadcrumb. Now it displays all the submodules of the project and the components from the `Project.Main` module. Updated the screencast.

farmaazon reviewed (approved):
Excellent job!

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/breadcrumbs-integration-182675703

vitvakatu committed:
Bump wasm size limit

farmaazon committed:
Merge remote-tracking branch 'origin/develop' into wip/vitvakatu/breadcrumbs-integration-182675703

farmaazon committed:
Merge remote-tracking branch 'origin/develop' into wip/vitvakatu/breadcrumbs-integration-182675703

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/breadcrumbs-integration-182675703

vitvakatu committed:
Fix unit tests for builder.rs
